### ISSUE TITLE
Create then delete within view when using middleware

### ISSUE THREAD

**Issue by d-mcc-af** [etianen/django-reversion#391]
Hey, thanks for your work. 

When using the reversion middleware, if I create and save an object within a view, then immediately delete it, I get the following stack.

```
Stacktrace (most recent call last):

  File "django/core/handlers/base.py", line 199, in get_response
    response = middleware_method(request, response)
  File "reversion/middleware.py", line 37, in process_response
    self._close_revision(request)
  File "reversion/middleware.py", line 28, in _close_revision
    revision_context_manager.end()
  File "reversion/revisions.py", line 176, in end
    in manager_context.items()
  File "reversion/revisions.py", line 175, in <genexpr>
    for obj, data
  File "reversion/revisions.py", line 616, in <lambda>
    version_data = lambda: adapter.get_version_data(instance, self._revision_context_manager._db)
  File "reversion/revisions.py", line 100, in get_version_data
    object_id_int = int(obj.pk)
```

**Comment by etianen** [comment]
Can you include the full stack trace? It looks like the bottom is missing.

**Comment by chrismcmacken** [comment]
I'm having the same issue, here's the full stack trace. It looks like @d-mcc-af is using the middleware, but I'm manually flagging the revisions in each method. The error doesn't occur every time, sometimes it works fine and other times it explodes with this error. 

Traceback:
File "/usr/local/lib/python2.7/site-packages/django/core/handlers/base.py" in get_response
  111.                     response = wrapped_callback(request, *callback_args, **callback_kwargs)
File "/usr/local/lib/python2.7/site-packages/django/views/decorators/csrf.py" in wrapped_view
  57.         return view_func(*args, **kwargs)
File "/usr/local/lib/python2.7/site-packages/rest_framework/viewsets.py" in view
  85.             return self.dispatch(request, *args, **kwargs)
File "/usr/local/lib/python2.7/site-packages/rest_framework/views.py" in dispatch
  407.             response = self.handle_exception(exc)
File "/usr/local/lib/python2.7/site-packages/rest_framework/views.py" in dispatch
  404.             response = handler(request, *args, **kwargs)
File "/usr/local/lib/python2.7/site-packages/django/db/transaction.py" in inner
  394.                 return func(*args, **kwargs)
File "/usr/local/lib/python2.7/site-packages/reversion/revisions.py" in do_revision_context
  305.                 return func(*args, **kwargs)
File "/usr/local/lib/python2.7/site-packages/reversion/revisions.py" in __exit__
  298.             self._context_manager.end()
File "/usr/local/lib/python2.7/site-packages/reversion/revisions.py" in end
  184.                                 in manager_context.items()
File "/usr/local/lib/python2.7/site-packages/reversion/revisions.py" in <genexpr>
  183.                                 for obj, data
File "/usr/local/lib/python2.7/site-packages/reversion/revisions.py" in <lambda>
  624.                 version_data = lambda: adapter.get_version_data(instance, self._revision_context_manager._db)
File "/usr/local/lib/python2.7/site-packages/reversion/revisions.py" in get_version_data
  108.             object_id_int = int(obj.pk)

Exception Type: TypeError at /api/removal-request/197/.json
Exception Value: int() argument must be a string or a number, not 'NoneType'

**Comment by podloucky-init** [comment]
django-reversion has been used in our projects for a while now. We just triggered the same error. It can be reproduced with the following quick&dirty TestCase:
(this is for `django-reversion==1.8.2`)

```
from django.contrib.auth.models import User
from django.test import TestCase
import reversion


class ReversionTest(TestCase):
    def test_delete_after_create_works(self):
        reversion.register(User)
        with reversion.create_revision():
            u = User.objects.create(username='popeyethesailorman')
            u.delete()
```

This will produce the following backtrace:
```
Traceback (most recent call last):
  File "/usr/local/home/podloucky/build/newoekotex/step/tests/test_reversion.py", line 11, in test_delete_after_create_works
    u.delete()
  File "/usr/local/home/podloucky/.virtualenvs/oekotex-master-development/lib/python2.7/site-packages/reversion/revisions.py", line 289, in __exit__
    self._context_manager.end()
  File "/usr/local/home/podloucky/.virtualenvs/oekotex-master-development/lib/python2.7/site-packages/reversion/revisions.py", line 175, in end
    in manager_context.items()
  File "/usr/local/home/podloucky/.virtualenvs/oekotex-master-development/lib/python2.7/site-packages/reversion/revisions.py", line 174, in <genexpr>
    for obj, data
  File "/usr/local/home/podloucky/.virtualenvs/oekotex-master-development/lib/python2.7/site-packages/reversion/revisions.py", line 594, in <lambda>
    version_data = lambda: adapter.get_version_data(instance, self._revision_context_manager._db)
  File "/usr/local/home/podloucky/.virtualenvs/oekotex-master-development/lib/python2.7/site-packages/reversion/revisions.py", line 99, in get_version_data
    object_id_int = int(obj.pk)
TypeError: int() argument must be a string or a number, not 'NoneType'
```

I guess it is not intended behavior as the docs would mention it otherwise.

**Comment by podloucky-init** [comment]
OK, I just found Issue #314 which basically says that you MUST NOT delete an object in the same revision context that you used for creating the object.
Also, the middleware has been more or less deprecated.

Thank you for this nice piece of software!

**Comment by etianen** [comment]
This is still an issue, really. It shouldn’t crash, certainly. I think the expected behaviour is one of:

a) The revision DOES NOT contain the model that was created and deleted.
b) The revision DOES contain the model that was created and deleted.
c) A more useful error is raised.

I think it’s probably a or b.

**Comment by podloucky-init** [comment]
Personally, I'd say it's a). Since a revision contains the changes that take you from state A to state B, it is not relevant whether a model was created and deleted within the same revision. It is not available in either state anyway.

As a side note, option c) would also be OK at least for the moment. That way people know what to look out for. And judging from the number of issues referencing this problem, it is not that big of a deal.

**Comment by etianen** [comment]
I’d agree that (a) is preferred, and it’s also the most feasible to implement. Not a high priority, sure, as creating and deleting a model in the same request is an unusual thing to do. But still, this needs to be fixed. Unless a pull request comes along, I’ll look into this when I have some more free time.

**Comment by etianen** [comment]
Fixed in 356b87569cfab70aa695a80cc6a9053dcf17ec0b

**Comment by podloucky-init** [comment]
I was working on that issue and have seen that it has already been fixed. Submitted the corresponding test, though.
